File: quant1x/runtime/scheduler.cpp
#include <quant1x/runtime/scheduler.h>

AsyncScheduler::AsyncScheduler(size_t thread_count) : pool_(thread_count), running_(true), next_id_(1) {
    spdlog::info("start scheduler...");
    scheduler_thread_ = std::thread([this] { scheduler_loop(); });
    spdlog::info("start scheduler...OK");
}

AsyncScheduler::~AsyncScheduler() {
    spdlog::warn("stop scheduler begin");
    stop();
    spdlog::warn("stop scheduler end");
}

runtime::task_id
AsyncScheduler::schedule_cron(const std::string &name, const std::string &cron_expr, std::function<void()> task) {
    std::lock_guard lock(mutex_);

    if (!running_) {
        throw std::runtime_error("schedule_cron called after scheduler stopped");
    }

    try {
        const auto id        = next_id_++;
        const auto cron      = cron::make_cron(cron_expr);
        const auto first_run = cron::cron_next(cron, Clock::now());
    cron_tasks_.emplace(id, CronTask{false, false, cron, std::move(task)});
        enqueue_task(ScheduledTask{first_run, [this, id, name] { execute_cron_task(id, name); }, id, name});
        ++st_scheduled_;

        return id;
    } catch (const cron::bad_cronexpr &e) {
        throw std::invalid_argument("Invalid cron expression: " + std::string(e.what()));
    }
}

void AsyncScheduler::cancel(runtime::task_id id) {
    std::lock_guard lock(mutex_);
    auto it = cron_tasks_.find(id);
    if (it != cron_tasks_.end()) {
        it->second.canceled = true; // 软取消，避免执行路径崩溃
        ++st_canceled_;
    }
    condition_.notify_all();
}

void AsyncScheduler::scheduler_loop() {
    spdlog::info("scheduler_loop...start");
    while (running_) {
        std::unique_lock lock(mutex_);

        if (task_queue_.empty()) {
            condition_.wait(lock, [this]{ return !task_queue_.empty() || !running_; });
            if (!running_) break;
            continue;
        }

        const auto &top_task = task_queue_.top();
        auto now = Clock::now();
        if (now < top_task.next_run) {
            condition_.wait_until(lock, top_task.next_run, [this]{ return !running_; });
            if (!running_) break;
            continue;
        }

        ScheduledTask task_to_run = top_task;
        task_queue_.pop();
        lock.unlock();

        if (!running_) break;
        // 如果是 cron 任务且已经标记取消, 不执行
        if (auto it = cron_tasks_.find(task_to_run.id); it != cron_tasks_.end() && it->second.canceled) {
            spdlog::debug("跳过取消任务 id={}, name={}", task_to_run.id, task_to_run.name);
            ++st_skipped_cancel_;
            continue;
        }

        pool_.detach_task([this, task_to_run] {
            try {
                if (running_) {
                    task_to_run.task();
                }
            } catch (const std::exception &e) {
                spdlog::error("任务执行异常 id={}, name={}, error={}", task_to_run.id, task_to_run.name, e.what());
            } catch (...) {
                spdlog::error("任务执行未知异常 id={}, name={}", task_to_run.id, task_to_run.name);
            }
        });
    }
    spdlog::info("scheduler_loop...stop");
}

void AsyncScheduler::reschedule_cron(runtime::task_id id, const std::string &name, const cron::cronexpr &cron) {
    std::lock_guard lock(mutex_);
    if (auto it = cron_tasks_.find(id); it == cron_tasks_.end() || it->second.canceled)
        return; // 已删除或已取消

    try {
        const auto next_time = cron::cron_next(cron, Clock::now());
        enqueue_task(ScheduledTask{next_time, [this, id, name] { execute_cron_task(id, name); }, id, name});
        ++st_rescheduled_;
    } catch (const cron::bad_cronexpr &e) {
        // 处理无效的cron表达式
        spdlog::error("Failed to reschedule cron task {}: {}", id, e.what());
    }
}

void AsyncScheduler::stop() {
    if (!running_.exchange(false)) {
        return;  // 确保只执行一次关闭
    }
    // 1. 唤醒调度线程使其尽快退出（不急于清容器，避免执行路径访问已清空结构）
    condition_.notify_all();

    // 2. 等待调度器线程结束
    if (scheduler_thread_.joinable()) {
        scheduler_thread_.join();
    }

    // 3. 等待线程池中已经派发的任务完成
    pool_.wait();

    // 4. 现在没有并发访问了，安全清理容器
    {
        std::lock_guard lock(mutex_);
        cron_tasks_.clear();
        while (!task_queue_.empty()) task_queue_.pop();
    }
}

void AsyncScheduler::execute_cron_task(runtime::task_id id, const std::string &name) {
    std::function<void()> task;
    cron::cronexpr        expr;
    bool                  need_reschedule = false;

    {
        std::lock_guard lock(mutex_);
        auto it = cron_tasks_.find(id);
        if (it == cron_tasks_.end()) return; // 已被stop清理或不存在
        auto &ct = it->second;
        if (ct.canceled) return;             // 已取消
        if (ct.cron_running) {
            spdlog::warn("Task {} skipped: previous execution still running", id);
            ++st_skipped_running_;
            return;
        }
        ct.cron_running = true;
        task            = ct.task;
        expr            = ct.expr;
        need_reschedule = true;              // 先假定要重排，再根据后续状态确认
    }

    try {
        task();
    } catch (const std::exception &e) {
        spdlog::error("execute_cron_task - 标准异常: {} (type: {})", e.what(), typeid(e).name());
    } catch (...) {
        spdlog::error("execute_cron_task - 未知异常 id={}", id);
    }

    {
        std::lock_guard lock(mutex_);
        auto it = cron_tasks_.find(id);
        if (it != cron_tasks_.end()) {
            it->second.cron_running = false;
            if (it->second.canceled || !running_) {
                need_reschedule = false;
            }
        } else {
            need_reschedule = false;
        }
    }

    if (need_reschedule) {
        ++st_executed_;
        reschedule_cron(id, name, expr);
    }
}

void AsyncScheduler::enqueue_task(AsyncScheduler::ScheduledTask &&task) {
    task_queue_.push(std::move(task));
    condition_.notify_all();
}

AsyncScheduler::Stats AsyncScheduler::get_stats() const {
    Stats s;
    s.scheduled       = st_scheduled_.load(std::memory_order_relaxed);
    s.executed        = st_executed_.load(std::memory_order_relaxed);
    s.skipped_cancel  = st_skipped_cancel_.load(std::memory_order_relaxed);
    s.skipped_running = st_skipped_running_.load(std::memory_order_relaxed);
    s.rescheduled     = st_rescheduled_.load(std::memory_order_relaxed);
    s.canceled        = st_canceled_.load(std::memory_order_relaxed);
    return s;
}

File: quant1x/runtime/scheduler.go
package runtime

import (
	"container/heap"
	"fmt"
	"runtime"
	"sync"
	"sync/atomic"
	"time"

	"github.com/robfig/cron/v3"
)

// AsyncScheduler provides asynchronous task scheduling with cron support.
// Tasks are delayed if the previous execution is still running.
type AsyncScheduler struct {
	pool          chan func() // worker pool channel
	running       int32       // atomic flag
	schedulerDone chan struct{}
	taskQueue     *taskHeap
	mutex         sync.Mutex
	cond          *sync.Cond
	nextID        int64 // atomic
	cronTasks     map[int64]*cronTask
	// stats
	stScheduled      int64
	stExecuted       int64
	stSkippedCancel  int64
	stSkippedRunning int64
	stRescheduled    int64
	stCanceled       int64
}

type cronTask struct {
	cronRunning bool
	canceled    bool
	expr        cron.Schedule
	task        func()
}

type scheduledTask struct {
	nextRun time.Time
	task    func()
	id      int64
	name    string
	index   int // for heap
}

// taskHeap implements heap.Interface
type taskHeap []*scheduledTask

func (h taskHeap) Len() int           { return len(h) }
func (h taskHeap) Less(i, j int) bool { return h[i].nextRun.Before(h[j].nextRun) }
func (h taskHeap) Swap(i, j int) {
	h[i], h[j] = h[j], h[i]
	h[i].index = i
	h[j].index = j
}
func (h *taskHeap) Push(x interface{}) {
	n := len(*h)
	item := x.(*scheduledTask)
	item.index = n
	*h = append(*h, item)
}
func (h *taskHeap) Pop() interface{} {
	old := *h
	n := len(old)
	item := old[n-1]
	old[n-1] = nil
	item.index = -1
	*h = old[0 : n-1]
	return item
}

// NewAsyncScheduler creates a new AsyncScheduler with the specified number of workers.
func NewAsyncScheduler(threadCount int) *AsyncScheduler {
	if threadCount <= 0 {
		threadCount = runtime.NumCPU()
	}
	s := &AsyncScheduler{
		pool:          make(chan func(), threadCount),
		running:       1,
		schedulerDone: make(chan struct{}),
		taskQueue:     &taskHeap{},
		cronTasks:     make(map[int64]*cronTask),
	}
	s.cond = sync.NewCond(&s.mutex)
	heap.Init(s.taskQueue)

	// start workers
	for i := 0; i < threadCount; i++ {
		go s.worker()
	}

	// start scheduler loop
	go s.schedulerLoop()

	return s
}

// ScheduleCron schedules a cron task.
func (s *AsyncScheduler) ScheduleCron(name, cronExpr string, task func()) (int64, error) {
	s.mutex.Lock()
	defer s.mutex.Unlock()

	if atomic.LoadInt32(&s.running) == 0 {
		return 0, fmt.Errorf("schedule_cron called after scheduler stopped")
	}

	expr, err := cron.ParseStandard(cronExpr)
	if err != nil {
		return 0, fmt.Errorf("invalid cron expression: %v", err)
	}

	id := atomic.AddInt64(&s.nextID, 1)
	firstRun := expr.Next(time.Now())
	s.cronTasks[id] = &cronTask{
		cronRunning: false,
		canceled:    false,
		expr:        expr,
		task:        task,
	}
	s.enqueueTask(&scheduledTask{
		nextRun: firstRun,
		task:    func() { s.executeCronTask(id, name) },
		id:      id,
		name:    name,
	})
	atomic.AddInt64(&s.stScheduled, 1)

	return id, nil
}

// Cancel cancels a scheduled task.
func (s *AsyncScheduler) Cancel(id int64) {
	s.mutex.Lock()
	defer s.mutex.Unlock()
	if ct, ok := s.cronTasks[id]; ok {
		ct.canceled = true
		atomic.AddInt64(&s.stCanceled, 1)
	}
	s.cond.Broadcast()
}

// Stop stops the scheduler.
func (s *AsyncScheduler) Stop() {
	if !atomic.CompareAndSwapInt32(&s.running, 1, 0) {
		return
	}
	s.cond.Broadcast()
	close(s.schedulerDone)
	// wait for workers to finish
	close(s.pool)
}

// Stats holds scheduler statistics.
type Stats struct {
	Scheduled      int64
	Executed       int64
	SkippedCancel  int64
	SkippedRunning int64
	Rescheduled    int64
	Canceled       int64
}

// GetStats returns current statistics.
func (s *AsyncScheduler) GetStats() Stats {
	return Stats{
		Scheduled:      atomic.LoadInt64(&s.stScheduled),
		Executed:       atomic.LoadInt64(&s.stExecuted),
		SkippedCancel:  atomic.LoadInt64(&s.stSkippedCancel),
		SkippedRunning: atomic.LoadInt64(&s.stSkippedRunning),
		Rescheduled:    atomic.LoadInt64(&s.stRescheduled),
		Canceled:       atomic.LoadInt64(&s.stCanceled),
	}
}

func (s *AsyncScheduler) enqueueTask(task *scheduledTask) {
	heap.Push(s.taskQueue, task)
	s.cond.Broadcast()
}

func (s *AsyncScheduler) executeCronTask(id int64, name string) {
	var task func()
	var expr cron.Schedule
	var needReschedule bool

	s.mutex.Lock()
	ct, ok := s.cronTasks[id]
	if !ok {
		s.mutex.Unlock()
		return
	}
	if ct.canceled {
		s.mutex.Unlock()
		return
	}
	if ct.cronRunning {
		atomic.AddInt64(&s.stSkippedRunning, 1)
		s.mutex.Unlock()
		return
	}
	ct.cronRunning = true
	task = ct.task
	expr = ct.expr
	needReschedule = true
	s.mutex.Unlock()

	// Reschedule immediately after setting running
	s.rescheduleCron(id, name, expr)

	// execute task
	defer func() {
		if r := recover(); r != nil {
			// log error
		}
	}()
	task()

	s.mutex.Lock()
	if ct, ok := s.cronTasks[id]; ok {
		ct.cronRunning = false
		if ct.canceled || atomic.LoadInt32(&s.running) == 0 {
			needReschedule = false
		}
	} else {
		needReschedule = false
	}
	s.mutex.Unlock()

	if needReschedule {
		atomic.AddInt64(&s.stExecuted, 1)
		// No need to reschedule again, already done at start
	}
}

func (s *AsyncScheduler) rescheduleCron(id int64, name string, expr cron.Schedule) {
	s.mutex.Lock()
	defer s.mutex.Unlock()
	if _, ok := s.cronTasks[id]; !ok || s.cronTasks[id].canceled {
		return
	}
	nextTime := expr.Next(time.Now())
	s.enqueueTask(&scheduledTask{
		nextRun: nextTime,
		task:    func() { s.executeCronTask(id, name) },
		id:      id,
		name:    name,
	})
	atomic.AddInt64(&s.stRescheduled, 1)
}

func (s *AsyncScheduler) schedulerLoop() {
	for atomic.LoadInt32(&s.running) == 1 {
		s.mutex.Lock()
		for s.taskQueue.Len() == 0 && atomic.LoadInt32(&s.running) == 1 {
			s.cond.Wait()
		}
		if atomic.LoadInt32(&s.running) == 0 {
			s.mutex.Unlock()
			break
		}
		task := heap.Pop(s.taskQueue).(*scheduledTask)
		now := time.Now()
		if now.Before(task.nextRun) {
			heap.Push(s.taskQueue, task)
			waitDur := task.nextRun.Sub(now)
			s.mutex.Unlock()
			select {
			case <-time.After(waitDur):
			case <-s.schedulerDone:
				return
			}
			continue
		}
		s.mutex.Unlock()

		if atomic.LoadInt32(&s.running) == 0 {
			break
		}
		// check cancel
		s.mutex.Lock()
		if ct, ok := s.cronTasks[task.id]; ok && ct.canceled {
			atomic.AddInt64(&s.stSkippedCancel, 1)
			s.mutex.Unlock()
			continue
		}
		s.mutex.Unlock()

		// submit to pool
		select {
		case s.pool <- task.task:
		case <-s.schedulerDone:
			return
		}
	}
}

func (s *AsyncScheduler) worker() {
	for {
		select {
		case task, ok := <-s.pool:
			if !ok {
				return
			}
			if atomic.LoadInt32(&s.running) == 1 {
				task()
			}
		case <-s.schedulerDone:
			return
		}
	}
}
